R/system.R

```diff
@@ -3290,3 +3290,135 @@ load_fred <- function(series_id, date_start = "", date_end = "", password) {
 }
 
 
+# Names of functions that uses column specifications, but would never reference outside data frame, such as select.
+# Collected from the doc of dplyr, tidyr, and our command menu.
+select_and_friends <- c('arrange', 'select', 'rename', 'relocate', 'reorder_cols',
+  'group_by', 'gather', 'spread', 'pivot_longer', 'pivot_wider', 'complete', 'expand', 'extract',
+  'unnest', 'unnest_longer', 'unnest_wider', 'nest', 'hoist', 'nest_legacy', 'unnest_legacy', 'pack', 'unpack',
+  'separate', 'unite', 'separate_rows', 'pivot',
+  'top_n',
+  'distinct',
+  'get_dupes',
+  'drop_na',
+  'fill',
+  'exp_balance',
+  'one_hot',
+  'do_cor',
+  'do_dist',
+  'do_cosine_sim.kv',
+  'do_survfit',
+  'do_market_impact',
+  'do_cmdscale',
+  'do_svd',
+  'exp_ts_cluster',
+  'do_anomaly_detection',
+  'exp_bayes_ab',
+  'evaluate_regression',
+  'evaluate_binary',
+  'evaluate_multi',
+  'do_roc',
+  'build_multinom',
+  'do_t.test',
+  'do_var.test',
+  'do_chisq.test',
+  'build_lm',
+  'build_lr',
+  'build_multinom',
+  'build_glm',
+  'build_model',
+  'build_coxph',
+  'do_tokenize',
+  'do_ngram',
+  'do_tf_idf',
+  'pair_count')
+# Names of functions that uses column specifications or reference with the column names, and could also reference outside data frames, such as mutate.
+# Collected from the doc of dplyr, and our command menu.
+mutate_and_friends <- c('mutate_group', 'mutate', 'mutate_at', 'mutate_all', 'mutate_if', 'transmute', 'summarize_group',
+  'summarize', 'summarize_at', 'summarize_all', 'summarize_if', 'summarise', 'summarise_at', 'summarise_all', 'summarise_if', 'filter',
+  'do_prophet') # do_prophet is here to handle reference to holiday data frame.
+
+# Returns names that references outside objects (most likely data frames) from the call.
+get_refs_in_call <- function(call,
+                             inside_mutate_and_friends = FALSE,
+                             inside_bang = FALSE,
+                             inside_bang_bang = FALSE) {
+  if (rlang::call_name(call) %in% select_and_friends) {
+    res <- c()
+  }
+  else {
+    # State transitions on inside_mutate_and_friends, inside_bang, and inside_bang_bang.
+    # It seems rlang::parse_expr does not recognize !! as one function, and rather recognize it as 2 separate bangs.
+    # So we need a state machine to detect it.
+    if (inside_mutate_and_friends) {
+      if (rlang::call_name(call) == '!') {
+        if (!inside_bang) {
+          inside_bang <- TRUE
+        }
+        else { # Already inside a bang.
+          inside_bang_bang <- TRUE
+        }
+      }
+      else {
```

Comment:
Please add comment for the else condition.

Reply:
Added some more comments.